### PR TITLE
Write .debug_frame

### PR DESCRIPTION
**Pull request by yurydelendik**
Basic functionality to write .debug_frame section as it is used at https://github.com/CraneStation/wasmtime/pull/53

/cc @bjorn3

### CODE REVIEW

File: src/common.rs [hunk 1]
```diff
@@ -188,3 +188,290 @@ pub enum UnitSectionOffset<T = usize> {
...
+    /// > The DW_CFA_set_loc instruction takes a single operand that represents
+    /// > a target address. The required action is to create a new table row
+    /// > using the specified address as the location. All other values in the
+    /// > new row are initially identical to the current row. The new location
+    /// > value is always greater than the current one. If the segment_size
+    /// > field of this FDE's CIE is non- zero, the initial location is preceded
+    /// > by a segment selector of the given length.
+    SetLoc {
+        /// The target address.
+        address: u64,
```
Comment (philipc) [inline src/common.rs]: Needs to be `write::Address` for writing.

File: src/write/frame.rs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        writer.write_u8(constants::DW_CFA_nop.0)?;
+    }
+    Ok(())
+}
+
+/// A `FrameDescriptionEntry` is a set of CFA instructions for an address range.
+#[derive(Debug)]
+pub struct FrameDescriptionEntry {
+    /// > The address of the first location associated with this table entry.
+    pub initial_location: u64,
```
Comment (philipc) [inline src/write/frame.rs]: Needs to be `write::Address`.

File: src/read/cfi.rs [hunk 3]
```diff
@@ -2798,298 +2798,11 @@ impl<R: Reader> RegisterRule<R> {
...
-    Nop,
-}
-
 const CFI_INSTRUCTION_HIGH_BITS_MASK: u8 = 0b1100_0000;
 const CFI_INSTRUCTION_LOW_BITS_MASK: u8 = !CFI_INSTRUCTION_HIGH_BITS_MASK;
 
-impl<R: Reader> CallFrameInstruction<R> {
-    fn parse(input: &mut R, address_size: u8) -> Result<CallFrameInstruction<R>> {
+impl<R: Reader> CallFrameInstruction<Expression<R>> {
+    fn parse(input: &mut R, address_size: u8) -> Result<CallFrameInstruction<Expression<R>>> {
```
Comment (philipc) [inline src/read/cfi.rs]: Use a type alias instead of changes like this.

**Comment by yurydelendik** [comment]
> 
Reusing `CallFrameInstruction` is great if it works well, but duplicating it is fine

Remove reusing, just duplicated the code.

**Comment by philipc** [comment]
@yurydelendik Are you still working on this? Would you like me to continue this work? I've been looking into CFI recently so I have a better idea of what's required here now. I'm happy to either review or take it up.

**Comment by yurydelendik** [comment]
> Are you still working on this? Would you like me to continue this work? I've been looking into CFI recently so I have a better idea of what's required here now. I'm happy to either review or take it up.

@philipc not at the moment, you can take it

**Comment by philipc** [comment]
Closed in favor of #412.

**Pull request closed without merging**